nixos/modules/services/security/vault.nix

```diff
@@ -0,0 +1,196 @@
+{ config, lib, pkgs, utils, ... }:
+
+with lib;
+let
+
+  inherit (pkgs) vault;
+
+  cfg = config.services.vault;
+ 
+  configFile = pkgs.writeText "vault.hcl" ''
+    listener "tcp" {
+      address = "${cfg.listener.address}"
+      ${if cfg.listener.cluster_address != "" then "cluster_address = \"${cfg.listener.cluster_address}\"" else ""  } 
+      
+      ${if cfg.listener.tls_cert_file != "" then "tls_cert_file = \"${cfg.listener.tls_cert_file}\"" else ""  } 
+      ${if cfg.listener.tls_key_file != "" then "tls_key_file = \"${cfg.listener.tls_key_file}\"" else "" } 
+      ${if cfg.listener.tls_disable != "false" then "tls_disable = \"1\"" else "" } 
+
+      ${if cfg.listener.tls_min_version != "tls12" then "tls_min_version = \"${cfg.listener.tls_min_version}\""  else ""} 
+
+      ${if cfg.listener.tls_cipher_suites != "" then "tls_cipher_suites = \"${cfg.listener.tls_cipher_suites}\"" else "" } 
+
+      ${if cfg.listener.tls_prefer_server_cipher_suites != "false" then "tls_prefer_server_cipher_suites = \"${cfg.listener.tls_prefer_server_cipher_suites}\"" else "" } 
+
+      ${if cfg.listener.tls_require_and_verify_client_cert != "false" then "tls_require_and_verify_client_cert= \"${cfg.listener.tls_require_and_verify_client_cert}\"" else "" } 
+
+    }
+
+    storage "${cfg.storage.backend}" {
+      ${cfg.storage.extraConfig}
+    }
+
+    ${if cfg.telemetry.extraConfig != "" then "
+      telemetry { 
+        ${if cfg.telemetry.disable_hostname then "disable_hostname = \"true\"" else ""}
+        ${cfg.telemetry.extraConfig}
+      }" else ""}
+
+  '';
+
+in
+{
+  options = {
+
+    services.vault = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Enables the vault daemon.
+        '';
+      };
+
+      listener = {
+
+        address = mkOption {
+          type = types.str;
+          default = "127.0.0.1:8200";
+          description = ''
+            The name of the ip interface to listen to.
+          '';
+        };
+
+        cluster_address = mkOption {
+          type = types.str;
```

Comment:
Rather than using `""` to denote an unset `cluster_address`, I'd recommend making this `types.nullOr types.str`, with `default = null`.